text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,657 @@
+the implicit nature of the duplication.
+
+Random number generators can implement `Clone` in a way that
+duplicates the state and causes identical/correlated sequences. In
+fact, having identical sequences is likely one of the most valuable
+uses of this, to e.g. replay/re-record failing test runs. (D's
+`std.random` offers `save` on many RNGs for this purpose, and C++'s
+`<random>` has copy constructors implemented for most types.)
+
+### Alternatives
+
+Provide a non-generic way to duplicate an RNG, e.g. a manual `fn
+save(&self)` on each type. This gets annoying if the RNG has to be
+used in generic code in any way, although it is somewhat questionable
+if generic code duplicating an RNG is correct (that said, it could
+just be code that is generic over different RNG types, e.g. a general
+framework for rerunning/recording tests as mentioned above).
+
+## `std::rand::distributions`
+
+Moved from the standard library into a crates.io package.
+
+## Omissions
+
+The following is a (non-exhaustive) list of things which are explicitly not included in
+this cut-down version of `std::rand`:
+
+- Generically handling RNGs more deeply, e.g.:
+    - Seeding
+    - Splitting RNGs to run in parallel, uncorrelatedly
+    - Fast-forwarding/rewinding (e.g. `ChaChaRng` provides
+      `set_counter` to jump around in the output sequence, and it can
+      be done efficiently with `XorShiftRng` via matrix exponentiation.)
+- Many common and historical RNG algorithms, e.g. MT19937, linear
+  congruence RNGs.
+- "RNG adaptors" like `discard_block_engine`, `independent_bits_engine`
+  in C++11's `random` header.
+- Detecting process forks and providing user-space RNGs the tools to
+  deal with it. A PRNG (almost always) stores its entire state in
+  memory, and a forked process will receive a copy of this state,
+  meaning the RNG in the main process and the one in the forked
+  process will generate the same sequence.
+- Non-uniform random numbers (e.g. what will have been `std::rand::distributions`)
+
+# Drawbacks
+
+- Losing functionality is sad.
+- `gen_iter(a..b)` (i.e. explicit end-points) will be surprisingly
+  inefficient for integers: to ensure uniformity and avoid bias and
+  maintain efficiency it has to approximately do rejection sampling
+  (after some precomputation). The author cannot see a way that this
+  can be addressed for the `..` notation specifically without
+  introducing more complexity (e.g. provide a way to generate a
+  "cached" version of the precomputation which can then be used for
+  generating every value in the iterator).
+- `gen_iter` taking by value may encourage users to call
+  `rng.clone().gen_iter()` instead of `(&mut rng).gen_iter()` if they
+  need it to not be consumed.
+
+# Alternatives
+
+Listed inline.
+
+# Unresolved questions
+
+- Both C++ and D use 'engine' instead of 'generator', but literature
+  generally seems to use 'generator'. Which term suits Rust best?
```

Comment:
I overcame my laziness and picked up my copy of Oxford dictionary. These are the definition of the relevant words:

* **engine** mechanical contrivance consisting of several parts working together, esp. as a source of power.
* **generator**
  1. a machine for converting mechanical into electrical energy;
  2. an apparatus for producing gas, steam, etc.

I also looked up: **generate** – bring into existence, produce, evolve.

---

So from these definition engine is 100% a wrong word for this, while squeezing our random bit producing algorithms into the “etc.” that goes after “gas and steam” sounds tenable.